### PR TITLE
e2e test perf improvements

### PR DESCRIPTION
**Pull request by sobolk**
## Problem

There are two inefficiencies in e2e tests addressed by this PR:
1. The async lock used in e2e tests, even on happy path, leaves timeout and a promise hanging around and blocking event loop from exiting. This is adding unnecessary wait when debugging locally while disabling cleanup portions of the test (I don't expect this is affecting CI, but local pain is enough to address it)
2. The `hot-swaps a change` is unnecessary burning 15+ seconds for tests that have no updates to test while doubling coverage with other steps.

## Changes

1. Handle timeout in async lock better
2. Conditionally run `hot-swaps a change` test logic.

**Corresponding docs PR, if applicable:**

## Checklist

- [ ] If this PR includes a functional change to the runtime behavior of the code, I have added or updated automated test coverage for this change.
- [ ] If this PR requires a change to the [Project Architecture README](../PROJECT_ARCHITECTURE.md), I have included that update in this PR.
- [ ] If this PR requires a docs update, I have linked to that docs PR above.
- [ ] If this PR modifies E2E tests, makes changes to resource provisioning, or makes SDK calls, I have run the PR checks with the `run-e2e` label set.

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: packages/integration-tests/src/async_lock.ts [hunk 1]
```diff
@@ -28,34 +34,45 @@ export class AsyncLock {
...
-    if (resolve) {
-      resolve();
+    const queueItem = this.queue.shift();
+    if (queueItem) {
+      queueItem.resolve();
+      if (queueItem.timeoutResolve) {
+        // if timeout was set resolve related promise and clear timeout
+        // so that it doesn't block node from exiting.
+        queueItem.timeoutResolve();
+        clearTimeout(queueItem.timeout);
```
Comment (sobolk) [inline packages/integration-tests/src/async_lock.ts:74]: This is kind of untestable. We're doing this to clear event loop earlier on happy case.

**Review by rtpascual — APPROVED** [review]

File: packages/integration-tests/src/test-e2e/deployment.test.ts [hunk 2]
```diff
@@ -157,28 +157,30 @@ void describe('deployment tests', { concurrency: testConcurrencyLevel }, () => {
...
-              }
-            );
-
             const updates = await testProject.getUpdates();
-            for (const update of updates) {
-              processController
-                .do(replaceFiles(update.replacements))
-                .do(ensureDeploymentTimeLessThan(update.deployThresholdSec));
-            }
+            if (updates.length > 0) {
```
Comment (0618) [inline packages/integration-tests/src/test-e2e/deployment.test.ts:161]: When can the `updates.length` not be greater `0`?

Comment (sobolk) [reply packages/integration-tests/src/test-e2e/deployment.test.ts:161]: Compare https://github.com/aws-amplify/amplify-backend/blob/320a86d053e5131b4c7dc45c7eff29d4e07ba3d6/packages/integration-tests/src/test-project-setup/data_storage_auth_with_triggers.ts#L174 
with
https://github.com/aws-amplify/amplify-backend/blob/320a86d053e5131b4c7dc45c7eff29d4e07ba3d6/packages/integration-tests/src/test-project-setup/test_project_base.ts#L122

In other words, only one project has any updates currently. Therefore, for projects without updates this test (in previous form) is 100% overlapping with "deploys fully" above.

Comment (sobolk) [reply packages/integration-tests/src/test-e2e/deployment.test.ts:161]: Also a search https://github.com/search?q=repo%3Aaws-amplify%2Famplify-backend%20%22async%20getUpdates()%22&type=code

**Review by 0618 — APPROVED** [review]

**Pull request merged**